### PR TITLE
Add coreDNS integrations 

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
### Summary
Add integration support for the coredns charm.

### Changes
- Extended charm functionality to handle events in case of DNS switch.
- Added unit and integration tests for coredns.
- Added the new tests in GitHub workflow.

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -42,3 +42,6 @@ requires:
     interface: microk8s-info
     scope: global
     limit: 1
+  dns:
```
Comment (addyess) [inline metadata.yaml:45]: I thought you said you wanted to change this to `dns-provider` to match `kubernetes-control-plane` and `coredns` charm

File: tests/integration/test_coredns.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        try:
+            LOG.info("Consume offer for dns-provider")
+            await e.model.consume(f"admin/{model_name}.coredns", "coredns")
+            LOG.info("Add relation between microk8s and coredns")
+            await e.model.add_relation("microk8s", "coredns")
+            LOG.info("Wait for idle")
+            await e.model.wait_for_idle(["microk8s"])
+
+            with e.model_context(k8s_model):
+                await e.model.wait_for_idle(["coredns"])
```
Comment (addyess) [inline tests/integration/test_coredns.py:54]: Perhaps there a test of some kind that coredns is handling dns requests?

File: tests/integration/test_coredns.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        finally:
+            app: Application = e.model.applications["microk8s"]
+            LOG.info("Remove relation between microk8s and coredns")
+            await app.remove_relation("microk8s:dns", "coredns:dns-provider")
```
Comment (addyess) [inline tests/integration/test_coredns.py:58]: I'm not sure the juju 3.1 cleaned up this model appropriately

```
ERROR    pytest_operator.plugin:plugin.py:875 ['cannot destroy application "coredns": application is used by 1 consumer']
Traceback (most recent call last):
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/pytest_operator/plugin.py", line 870, in _destroy
    await entity.destroy(**kwargs)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/application.py", line 249, in destroy
    res = await app_facade.DestroyApplication(applications=[client.DestroyApplicationParams(
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/facade.py", line 483, in wrapper
    reply = await f(*args, **kwargs)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/_client17.py", line 1074, in DestroyApplication
    reply = await self.rpc(msg)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/facade.py", line 656, in rpc
    result = await self.connection.rpc(msg, encoder=TypeEncoder)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/connection.py", line 661, in rpc
    raise errors.JujuError(err_results)
juju.errors.JujuError: ['cannot destroy application "coredns": application is used by 1 consumer']
```


```suggestion
            await app.remove_relation("dns", "coredns")
```

Comment (sachinkumarsingh092) [reply tests/integration/test_coredns.py:58]: Hey @addyess, this is something that keeps coming. I've tried locally by removing the relation (and the correct names) and even without removing relations. But during the resetting, this pops up.

File: metadata.yaml (hunk 1 above)
Comment (sachinkumarsingh092) [reply metadata.yaml:45]: Yeah, that was breaking things for some reason. I'll update these once everything works fine.

File: tests/integration/test_coredns.py (hunk 2 above)
Comment (sachinkumarsingh092) [reply tests/integration/test_coredns.py:54]: Added this.

File: tests/integration/test_coredns.py (hunk 3 above)
Comment (addyess) [reply tests/integration/test_coredns.py]: I've raised https://github.com/canonical/charm-microk8s/pull/100 into this branch which seems to have removed the errors in the tests i spotted a few days ago

File: metadata.yaml (hunk 1 above)
Comment (berkayoz) [reply metadata.yaml:45]: @sachinkumarsingh092 Any updates on this?

File: tests/integration/test_coredns.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        await microk8s.remove_relation("dns", "coredns")
+        await asyncio.gather(
+            machine_model.remove_saas("coredns"), machine_model.wait_for_idle(["microk8s"])
+        )
+
+
+@pytest.mark.abort_on_fail
+async def test_core_dns(related_microk8s_coredns):
+    microk8s, _ = related_microk8s_coredns
+    DNS_TEST = "microk8s kubectl run -i --tty --rm debug --image=busybox --restart=Never -- nslookup google.com"
```
Comment (berkayoz) [inline tests/integration/test_coredns.py:82]: I've seen weird DNS related issues in busybox images over the time, maybe a different image like ubuntu would be better? I'm hesitant on using busybox for anything DNS related.

**Review by berkayoz — COMMENTED** [review]
Looks good overall, left small comments.

File: tests/integration/test_coredns.py (hunk 4 above)
Comment (sachinkumarsingh092) [reply tests/integration/test_coredns.py:82]: I guess fixing to a particular version should be okay?

**Review by sachinkumarsingh092 — COMMENTED** [review]
LGTM to the changes.

**Pull request merged**